rfcs/0008-transition-frontier-controller.md

```diff
@@ -0,0 +1,337 @@
+
+3. The `Ktree.t` is mutable and exposes an $O(1)$ `lookup : State_hash.t -> node_entry` where `node_entry` contains (in our case) a `Breadcrumb.t`. We will discuss this further in [Ktree](#ktree).
+
+Rationale: We perform `lookup` and `add` frequently on this structure and `add` can be $O(1)$ in the presence of this `lookup` too. `lookup` also lets us traverse the tree forwards and backwards without explicit backedges in our [rose tree](https://en.wikipedia.org/wiki/Rose_tree) -- further simplifying the `Ktree.t` implementation.
+
+<a href="transition-handler"></a>
+### Transition Handler
+
+The transition handler is broken down into two main components: Validator and
+Processor
+
+#### Validator
+
+* Input: `External_transition.t`
+* Output: `External_transition.t` (we should consider making an `External_transition.Validated.t` for output here)
+
+The validator receives as input external transitions from the network and
+performs a series of checks on the transition before passing it off to be
+processed. For simplicity it can also receive transitions created by the local
+proposer. In the future, we can consider skipping this step.
+
+Particularly, we validate transitions with the following checks (in this order):
+
+1. Checking for existence in the breadcrumb tree (we've already seen it)
+2. Checking for consensus errors
+3. Verifying the included SNARK
+
+This order is chosen because it's cheaper computationally to perform (1) and
+(2). So if we get lucky we can short-circuit before getting to the SNARK step.
+
+#### Processor
+
+* Input: `External_transition.t` (from validator) and `Breadcrumb.t` (from catchup)
+* Outputs: `External_transition.t` (to ledger catchup); modifying transition frontier
+
+The processor receives a single validated external transitions from the
+validator and then attempts to add the transition to the breadcrumb tree. The
+processor is the only "thread" allowed to write changes to the
+[Transition Frontier](#transition-frontier). All other threads must delegate to
+the processor.
+
+[Ledger Catchup](#ledger-catchup) needs to add a batch of transitions all at
+once, so it shares the add section and constructs the underlying breadcrumb wired up that can just be $O(1)$ added to the frontier tree.
+
+#### Adding to Transition Frontier
+
+The add process runs in two phases:
+
+1. Perform a `lookup` on the `Transition_frontier` for the previous `State_hash.t` of this transition. If it is absent, send to [catchup monitor](#catchup-monitor). If present, continue.
+2. Apply the `Ledger_builder_diff.t` to the `Breadcrumb.t` retrieved from (1). Note that this no longer mutates the underlying ledger-builder but rather constructs a new mask on top of the existing mask -- see [Transition Frontier](#transition-frontier) for more.
```

Comment:
I think this is better expressed as "derive a mask from the parent retrieved in (1) and apply the ledger builder diff of the transaction to that new mask".